File: src/main/java/ordilov/lissn/security/userinfo/UserPrincipal.java
package ordilov.lissn.security.userinfo;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import lombok.AllArgsConstructor;
import lombok.Getter;
import ordilov.lissn.member.domain.AuthInfo.TokenInfo;
import ordilov.lissn.member.domain.AuthProvider;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.core.user.OAuth2User;

@Getter
@AllArgsConstructor
public class UserPrincipal implements OAuth2User {
  private final Long id;
  private final AuthProvider provider;
  private final Map<String, Object> attributes;
  private final Collection<? extends GrantedAuthority> authorities;

  public UserPrincipal(TokenInfo tokenInfo) {
    ObjectMapper mapper = new ObjectMapper();
    this.id = tokenInfo.getId();
    this.provider = tokenInfo.getProvider();
    this.authorities = Collections.singleton(new SimpleGrantedAuthority("ROLE_USER"));
    this.attributes = mapper.convertValue(tokenInfo, Map.class);
  }

  @Override
  public Map<String, Object> getAttributes() {
    return attributes;
  }

  @Override
  public Collection<? extends GrantedAuthority> getAuthorities() {
    return authorities;
  }

  @Override
  public String getName() {
    return String.valueOf(id);
  }

  public TokenInfo getTokenInfo() {
    return new TokenInfo(id,
        attributes.get("name").toString(),
        attributes.get("email").toString(),
        attributes.get("picture").toString(),
        provider);
  }
}

File: src/main/java/ordilov/lissn/security/userinfo/UserPrincipal.kt
package ordilov.lissn.security.userinfo;

import com.fasterxml.jackson.databind.ObjectMapper
import ordilov.lissn.member.domain.AuthProvider
import ordilov.lissn.member.domain.TokenInfo
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.core.user.OAuth2User
import java.util.*

open class UserPrincipal(
    val id: Long,
    val provider: AuthProvider,
    attributes: MutableMap<String, Any>,
    authorities: MutableCollection<out GrantedAuthority>
) : OAuth2User {

    val attributes: MutableMap<String, Any> = attributes
        @JvmName("getAttributes_") get
    val authorities: MutableCollection<out GrantedAuthority> = authorities
        @JvmName("getAuthorities_") get

    constructor(tokenInfo: TokenInfo) : this(
        id = tokenInfo.id,
        provider = tokenInfo.provider,
        ObjectMapper().convertValue(tokenInfo, Map::class.java) as MutableMap<String, Any>,
        Collections.singleton(SimpleGrantedAuthority("ROLE_USER"))
    )

    override fun getName(): String {
        return id.toString()
    }

    override fun getAttributes(): MutableMap<String, Any> {
        return attributes
    }

    override fun getAuthorities(): MutableCollection<out GrantedAuthority> {
        return authorities
    }

    fun getTokenInfo(): TokenInfo {
        return TokenInfo(
            id,
            attributes["name"].toString(),
            attributes["email"].toString(),
            attributes["picture"].toString(),
            provider
        )
    }
}